src/ray/raylet/pubsub/pubsub_client.h

```diff
@@ -0,0 +1,162 @@
+                         const ObjectID &object_id) override;
+
+ private:
+  /// Encapsulates the subscription information such as subscribed object ids ->
+  /// callbacks.
+  /// TODO(sang): Use a channel abstraction instead of owner address once OBOD uses the
+  /// pubsub.
+  struct SubscriptionInfo {
+    SubscriptionInfo(const rpc::Address &pubsub_server_address)
+        : pubsub_server_address_(pubsub_server_address) {}
+
+    /// Address of the pubsub server.
+    const rpc::Address pubsub_server_address_;
+    // Object ID -> subscription_callback
+    absl::flat_hash_map<const ObjectID,
+                        std::pair<SubscriptionCallback, SubscriptionFailureCallback>>
+        subscription_callback_map_;
+  };
+
+  /// Create a long polling connection to the owner for receiving the published messages.
+  /// TODO(sang): Currently, we assume that unregistered objects will never be published
+  /// from the pubsub server. We may want to loose the restriction once OBOD is supported
+  /// by this function.
+  /// \param owner_address The address of the owner that publishes
+  /// objects.
+  /// \param subscriber_address The address of the subscriber.
+  void MakeLongPollingPubsubConnection(const rpc::Address &owner_address,
+                                       const rpc::Address &subscriber_address);
+
+  /// Private method to handle long polling responses. Long polling responses contain the
+  /// published messages.
+  void HandleLongPollingResponse(const rpc::Address &owner_address,
+                                 const rpc::Address &subscriber_address, Status &status,
+                                 const rpc::PubsubLongPollingReply &reply);
+
+  /// Private method to handle pubsub server failures.
+  void HandleCoordinatorFailure(const rpc::Address &owner_address);
+
+  /// Returns a subscription callback; The first entry indicates if the subscription
+  /// callback exists. The second entry is the callback itself.
+  inline std::pair<bool, SubscriptionCallback> GetSubscriptionCallback(
+      const rpc::Address &owner_address, const ObjectID &object_id) {
+    const auto owner_worker_id = WorkerID::FromBinary(owner_address.worker_id());
+    auto subscription_it = subscription_map_.find(owner_worker_id);
+    if (subscription_it == subscription_map_.end()) {
+      return std::make_pair(false, nullptr);
+    }
+    auto callback_it = subscription_it->second.subscription_callback_map_.find(object_id);
+    bool exist = callback_it != subscription_it->second.subscription_callback_map_.end();
+    if (!exist) {
+      return std::make_pair(false, nullptr);
+    }
+    auto &subscription_callback = callback_it->second.first;
+    return std::make_pair(true, subscription_callback);
+  }
+
+  /// Returns a owner failure callback; The first entry indicates if the subscription
+  /// callback exists. The second entry is the callback itself.
+  inline std::pair<bool, SubscriptionCallback> GetFailureCallback(
```

Comment:
Btw, I'm wondering why you put it in the class declaration? Usually, we put small functions(to make them inline for performance) or template functions in the body. Otherwise, we follow the style of the code.

Reply:
I just followed from other parts of code. Is your suggestion to move this into .cc without inline keywords?

Reply:
I'm good with either: remove inline here since it's inlined by default; or move it to .cc file.